### PR TITLE
`Development`: Refactor RepositoryAccessService and plagiarism access check

### PR DESCRIPTION
**Pull request by mateusmm01**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I documented the TypeScript code using JSDoc style.
- [x] I translated all newly inserted strings into English and German.

#### Changes affecting Programming Exercises
- [ ] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).
- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server configured with **Gitlab** and **Jenkins**.
- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server configured with **Bitbucket** and **Bamboo**.

### Motivation and Context

Currently, the `RepositoryAccessService` isn't very well written or easily readable. It plays a vital role in the security of Artemis programming exercises. Therefore, it should be refactored following these rules:
https://docs.artemis.cit.tum.de/admin/accessRights.html#programming-exercises-repository-access

### Description

`RepositoryAccessService` was refactored. We also fetch the files for the plagiarism check separately with their own endpoints. These check the plagiarism access instead of the general repository access.

### Steps for Testing

Prerequisites:
- 1 Admin
- 1 Instructor
- 1 Tutor
- 1 Student

- 1 Programming Exercises

Create different Programming Exercises with different rules and check that you can only access the repositories in the expected way (https://docs.artemis.cit.tum.de/admin/accessRights.html#programming-exercises-repository-access)

Also check that instructors and students can still view the plagiarism views as expected

#### Exam Mode Testing

Prerequisites:
- 1 Admin
- 1 Instructor
- 1 Tutor
- 1 Student

- 1 Exam with Programming Exercises

Create an Exam with different Programming Exercises with different rules and check that you can only access the repositories in the expected way (https://docs.artemis.cit.tum.de/admin/accessRights.html#programming-exercises-repository-access)

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2

## Summary by CodeRabbit

- **New Features**
	- Introduced a method to verify access to plagiarism submissions.
	- New endpoints to retrieve files for plagiarism view in programming exercises.
- **Refactor**
	- Simplified the constructor in `RepositoryAccessService`.
	- Enhanced access control logic in `RepositoryAccessService` to handle user roles and actions more effectively.
	- Improved file retrieval process in `RepositoryService` and `RepositoryResource` for better readability and maintainability.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/RepositoryAccessService.java [hunk 1]
```diff
@@ -59,159 +41,89 @@ public RepositoryAccessService(ParticipationAuthorizationCheckService participat
...
-                throw new IllegalStateException("The participation is locked but the reason is unknown.");
-            }
-
-            throw new AccessForbiddenException(errorMessage);
+        if (hasStarted) {
+            return hasAccessAfterExerciseStart(studentParticipation, repositoryActionType, isLocked);
         }
+        // Only teaching assistants have access to the repository before the exercise has started.
+        else
+            return isTeachingAssistant;
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/service/RepositoryAccessService.java:89]:
```suggestion
        if (!hasStarted) {
            // Only teaching assistants have access to the repository before the exercise has started.
            return isTeachingAssistang;
        }

        return hasAccessAfterExerciseStart(studentParticipation, repositoryActionType, isLocked);
```

File: src/main/java/de/tum/in/www1/artemis/service/RepositoryAccessService.java [hunk 2]
```diff
@@ -59,159 +41,89 @@ public RepositoryAccessService(ParticipationAuthorizationCheckService participat
...
-                throw new AccessForbiddenException();
-            }
-        }
-        else if (!hasAccessToSubmission && hasPlagiarismComparison) {
-            throw new AccessForbiddenException();
-        }
+    private boolean hasAccessAfterExerciseStart(StudentParticipation studentParticipation, RepositoryActionType repositoryActionType, boolean isLocked) {
+        // The user always has read permissions after the exercise has started.
+        if (repositoryActionType == RepositoryActionType.READ)
+            return true;
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/service/RepositoryAccessService.java:102]:
```suggestion
        if (repositoryActionType == RepositoryActionType.READ) {
            return true;
        }
```

File: src/main/java/de/tum/in/www1/artemis/service/RepositoryAccessService.java [hunk 3]
```diff
@@ -59,159 +41,89 @@ public RepositoryAccessService(ParticipationAuthorizationCheckService participat
...
+
+        // Check if the user has write or reset permissions.
+        else
+            return hasWriteOrResetPermissions(studentParticipation, isLocked);
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/service/RepositoryAccessService.java:106]:
```suggestion
        // Check if the user has write or reset permissions.
        return hasWriteOrResetPermissions(studentParticipation, isLocked);
```

File: src/main/java/de/tum/in/www1/artemis/service/RepositoryAccessService.java [hunk 4]
```diff
@@ -59,159 +41,89 @@ public RepositoryAccessService(ParticipationAuthorizationCheckService participat
...
-            // Retrieving the student via this type cast is safe because isOwner is false here which means that "programmingExerciseParticipation instanceof StudentParticipation"
-            // must have been true in the check above.
-            var optStudent = ((StudentParticipation) programmingExerciseParticipation).getStudent();
-            if (optStudent.isPresent() && programmingExercise.isExamExercise() && examSubmissionService.isAllowedToSubmitDuringExam(programmingExercise, optStudent.get(), false)) {
-                throw new AccessForbiddenException();
-            }
-        }
+        // The user has write or reset permissions if the participation is not locked and the due date has not passed yet.
+        if (beforeDueDate && !isLocked)
+            return true;
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/service/RepositoryAccessService.java:123]:
```suggestion
        if (beforeDueDate && !isLocked) {
            return true;
        }
```

File: src/main/java/de/tum/in/www1/artemis/service/RepositoryAccessService.java [hunk 5]
```diff
@@ -59,159 +41,89 @@ public RepositoryAccessService(ParticipationAuthorizationCheckService participat
...
+
+        // The user has write or reset permissions if due date has passed, but the participation is in practice mode.
+        return !beforeDueDate && isPracticeMode;
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/service/RepositoryAccessService.java:126]: Small technicality here:
Even if it's practice mode, you shouldn't have access to a locked repo. I assume practice mode and locked state won't ever occur at the same time, but it should still be correctly implemented.

File: src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryProgrammingExerciseParticipationResource.java [hunk 6]
```diff
@@ -92,17 +101,41 @@ Repository getRepository(Long participationId, RepositoryActionType repositoryAc
...
+
+    /**
+     * Get the repository for the plagiarism view of the given participation.
+     *
+     * @param participationId the id of the participation to retrieve the repository for
+     * @return the repository for the plagiarism view of the given participation
+     * @throws GitAPIException          if the repository could not be accessed
+     * @throws AccessForbiddenException if the user does not have access to the repository
+     */
+    Repository getRepositoryForPlagiarismView(Long participationId) throws GitAPIException, AccessForbiddenException {
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryProgrammingExerciseParticipationResource.java:118]: Looks like methods like these should rather be on the service layer.

File: src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryResource.java [hunk 7]
```diff
@@ -147,16 +147,28 @@ public ResponseEntity<byte[]> getFile(Long domainId, String filename) {
...
 
+    /**
+     * Helper method to get the content of a file from the given repository.
+     *
+     * @param filename   of the file to retrieve.
+     * @param repository the repository to retrieve the file from.
+     * @return the file if available.
+     * @throws IOException if the file can't be retrieved.
+     */
+    protected ResponseEntity<byte[]> getFileFromRepository(String filename, Repository repository) throws IOException {
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryResource.java:162]: Same here, would expect this to be part of a service

**Review by julian-christl — CHANGES_REQUESTED** [review]
Good job in general, I have a few remarks

File: src/main/java/de/tum/in/www1/artemis/service/RepositoryAccessService.java [hunk 8]
```diff
@@ -59,159 +41,91 @@ public RepositoryAccessService(ParticipationAuthorizationCheckService participat
...
-            // For an exam exercise, the student must not be allowed to submit anymore.
-            // Retrieving the student via this type cast is safe because isOwner is false here which means that "programmingExerciseParticipation instanceof StudentParticipation"
-            // must have been true in the check above.
-            var optStudent = ((StudentParticipation) programmingExerciseParticipation).getStudent();
-            if (optStudent.isPresent() && programmingExercise.isExamExercise() && examSubmissionService.isAllowedToSubmitDuringExam(programmingExercise, optStudent.get(), false)) {
-                throw new AccessForbiddenException();
-            }
-        }
+        // The user has write or reset permissions if due date has passed, but the participation is in practice mode and not locked.
+        return !beforeDueDate && isPracticeMode && !isLocked;
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/service/RepositoryAccessService.java:128]: We can actually return `false` for `isLocked == true` even before the initialization of the first variables.

**Review by julian-christl — DISMISSED** [review]
That's also a nice way of doing it. Good job with the changes! Code looks good now!

File: src/main/java/de/tum/in/www1/artemis/web/rest/TextSubmissionResource.java [hunk 9]
```diff
@@ -154,7 +154,7 @@ public ResponseEntity<TextSubmission> getTextSubmissionWithResults(@PathVariable
             // anonymize and throw exception if not authorized to view submission
             plagiarismService.checkAccessAndAnonymizeSubmissionForStudent(textSubmission, userRepository.getUser().getLogin(),
                     // TODO: might be better to use the individual due date here if available
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/TextSubmissionResource.java:156]: This is now integrated by using the exerciseDateService, right?
```suggestion
```

**Review by Strohgelaender — APPROVED** [review]
code looks good to me.

Tested locally, students are able to see the other student's code (in case of a plagriarism accousation) only after the deadline.

**Review by milljoniaer — APPROVED** [review]
Code LGTM

**Review by laurenzfb — APPROVED** [review]
demonstrated in testing session. lgtm

**Review by julian-christl — APPROVED** [review]
Reapprove code, changes look good

**Review by Strohgelaender — APPROVED** [review]

**Comment by krusche** [comment]
10 e2e tests consistently fail. Please double check what is happening there

**Pull request merged**